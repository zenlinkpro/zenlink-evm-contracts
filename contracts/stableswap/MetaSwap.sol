// SPDX-License-Identifier: MIT
pragma solidity >=0.8.0;

import { 
    StableSwap, 
    StableSwapStorage, 
    SafeERC20,
    IERC20,
    IStableSwap,
    LPToken
} from "./StableSwap.sol";
import { MetaSwapStorage } from "./MetaSwapStorage.sol";

contract MetaSwap is StableSwap {
    using MetaSwapStorage for StableSwapStorage.SwapStorage;
    using SafeERC20 for IERC20;

    MetaSwapStorage.MetaSwap public metaSwapStorage;

    // events replicated from SwapStorage to make the ABI easier for dumb
    // clients
    event TokenSwapUnderlying(
        address indexed buyer,
        uint256 tokensSold,
        uint256 tokensBought,
        uint128 soldId,
        uint128 boughtId
    );

    /**
     * @notice Get the virtual price, to help calculate profit
     * @return the virtual price, scaled to the POOL_PRECISION_DECIMALS
     */
    function getVirtualPrice()
        external
        view
        virtual
        override
        returns (uint256)
    {
        return MetaSwapStorage.getVirtualPrice(swapStorage, metaSwapStorage);
    }

    /**
     * @notice Calculate amount of tokens you receive on swap
     * @param tokenIndexFrom the token the user wants to sell
     * @param tokenIndexTo the token the user wants to buy
     * @param dx the amount of tokens the user wants to sell. If the token charges
     * a fee on transfers, use the amount that gets transferred after the fee.
     * @return amount of tokens the user will receive
     */
    function calculateSwap(
        uint8 tokenIndexFrom,
        uint8 tokenIndexTo,
        uint256 dx
    ) virtual external view override returns (uint256) {
        return MetaSwapStorage.calculateSwap(
            swapStorage,
            metaSwapStorage,
            tokenIndexFrom,
            tokenIndexTo,
            dx
        );
    }

    /**
     * @notice Calculate amount of tokens you receive on swap. For this function,
     * the token indices are flattened out so that underlying tokens are represented.
     * @param tokenIndexFrom the token the user wants to sell
     * @param tokenIndexTo the token the user wants to buy
     * @param dx the amount of tokens the user wants to sell. If the token charges
     * a fee on transfers, use the amount that gets transferred after the fee.
     * @return amount of tokens the user will receive
     */
    function calculateSwapUnderlying(
        uint8 tokenIndexFrom,
        uint8 tokenIndexTo,
        uint256 dx
    ) external view virtual returns (uint256) {
        return MetaSwapStorage.calculateSwapUnderlying(
            swapStorage,
            metaSwapStorage,
            tokenIndexFrom,
            tokenIndexTo,
            dx
        );
    }

    /**
     * @notice A simple method to calculate prices from deposits or
     * withdrawals, excluding fees but including slippage. This is
     * helpful as an input into the various "min" parameters on calls
     * to fight front-running
     *
     * @dev This shouldn't be used outside frontends for user estimates.
     *
     * @param amounts an array of token amounts to deposit or withdrawal,
     * corresponding to pooledTokens. The amount should be in each
     * pooled token's native precision. If a token charges a fee on transfers,
     * use the amount that gets transferred after the fee.
     * @param deposit whether this is a deposit or a withdrawal
     * @return token amount the user will receive
     */
    function calculateTokenAmount(uint256[] calldata amounts, bool deposit)
        virtual
        external
        view
        override
        returns (uint256)
    {
        return MetaSwapStorage.calculateTokenAmount(
            swapStorage,
            metaSwapStorage,
            amounts,
            deposit
        );
    }

    function calculateRemoveLiquidity(uint256 amount) virtual external view override returns (uint256[] memory) {
        return MetaSwapStorage.calculateRemoveLiquidity(swapStorage, amount);
    }

    /**
     * @notice Calculate the amount of underlying token available to withdraw
     * when withdrawing via only single token
     * @param tokenAmount the amount of LP token to burn
     * @param tokenIndex index of which token will be withdrawn
     * @return availableTokenAmount calculated amount of underlying token
     * available to withdraw
     */
    function calculateRemoveLiquidityOneToken(
        uint256 tokenAmount,
        uint8 tokenIndex
    ) virtual external view override returns (uint256) {
        return MetaSwapStorage.calculateRemoveLiquidityOneToken(
            swapStorage,
            metaSwapStorage,
            tokenAmount,
            tokenIndex
        );
    }

    function initialize(
        address[] memory,
        uint8[] memory,
        string memory,
        string memory,
        uint256,
        uint256,
        uint256,
        address
    ) public virtual override onlyAdmin {
        revert("use initializeMetaSwap() instead");
    }

    /**
     * @notice Initializes this MetaSwap contract with the given parameters.
     * MetaSwap uses an existing Swap pool to expand the available liquidity.
     * _pooledTokens array should contain the base Swap pool's LP token as
     * the last element. For example, if there is a Swap pool consisting of
     * [DAI, USDC, USDT]. Then a MetaSwap pool can be created with [sUSD, BaseSwapLPToken]
     * as _pooledTokens.
     *
     * This will also deploy the LPToken that represents users'
     * LP position. The owner of LPToken will be this contract - which means
     * only this contract is allowed to mint new tokens.
     *
     * @param _pooledTokens an array of ERC20s this pool will accept. The last
     * element must be an existing Swap pool's LP token's address.
     * @param decimals the decimals to use for each pooled token,
     * eg 8 for WBTC. Cannot be larger than POOL_PRECISION_DECIMALS
     * @param lpTokenName the long-form name of the token to be deployed
     * @param lpTokenSymbol the short symbol for the token to be deployed
     * @param _a the amplification coefficient * n * (n - 1). See the
     * StableSwap paper for details
     * @param _fee default swap fee to be initialized with
     * @param _adminFee default adminFee to be initialized with
     */
    function initializeMetaSwap(
        address[] memory _pooledTokens,
        uint8[] memory decimals,
        string memory lpTokenName,
        string memory lpTokenSymbol,
        uint256 _a,
        uint256 _fee,
        uint256 _adminFee,
        address _feeDistributor,
        IStableSwap baseSwap
    ) public virtual onlyAdmin {
        StableSwap.initialize(
            _pooledTokens, 
            decimals, 
            lpTokenName, 
            lpTokenSymbol, 
            _a, 
            _fee, 
            _adminFee, 
            _feeDistributor
        );

        metaSwapStorage.baseSwap = baseSwap;
        metaSwapStorage.baseVirtualPrice = baseSwap.getVirtualPrice();
        metaSwapStorage.baseCacheLastUpdated = block.timestamp;

        // Read all tokens that belong to baseSwap
        {
            uint8 i;
            for (; i < 32; i++) {
                try baseSwap.getToken(i) returns (IERC20 token) {
                    metaSwapStorage.baseTokens.push(token);
                    token.safeApprove(address(baseSwap), type(uint256).max);
                } catch {
                    break;
                }
            }
            require(i > 1, "baseSwap must pool at least 2 tokens");
        }

        // Check the last element of _pooledTokens is owned by baseSwap
        IERC20 baseLPToken = IERC20(_pooledTokens[_pooledTokens.length - 1]);
        require(
            LPToken(address(baseLPToken)).owner() == address(baseSwap),
            "baseLPToken is not owned by baseSwap"
        );

        // Pre-approve the baseLPToken to be used by baseSwap
        baseLPToken.safeApprove(address(baseSwap), type(uint256).max);
    }

    /**
     * @notice Swap two tokens using this pool
     * @param tokenIndexFrom the token the user wants to swap from
     * @param tokenIndexTo the token the user wants to swap to
     * @param dx the amount of tokens the user wants to swap from
     * @param minDy the min amount the user would like to receive, or revert.
     * @param deadline latest timestamp to accept this transaction
     */
    function swap(
        uint8 tokenIndexFrom,
        uint8 tokenIndexTo,
        uint256 dx,
        uint256 minDy,
        uint256 deadline
    ) external virtual override nonReentrant whenNotPaused deadlineCheck(deadline) returns (uint256) {
        return MetaSwapStorage.swap(
            swapStorage,
            metaSwapStorage,
            tokenIndexFrom,
            tokenIndexTo,
            dx,
            minDy
        );
    }

    /**
     * @notice Swap two tokens using this pool and the base pool.
     * @param tokenIndexFrom the token the user wants to swap from
     * @param tokenIndexTo the token the user wants to swap to
     * @param dx the amount of tokens the user wants to swap from
     * @param minDy the min amount the user would like to receive, or revert.
     * @param deadline latest timestamp to accept this transaction
     */
    function swapUnderlying(
        uint8 tokenIndexFrom,
        uint8 tokenIndexTo,
        uint256 dx,
        uint256 minDy,
        uint256 deadline
    ) external virtual nonReentrant whenNotPaused deadlineCheck(deadline) returns (uint256) {
        return MetaSwapStorage.swapUnderlying(
            swapStorage,
            metaSwapStorage,
            tokenIndexFrom,
            tokenIndexTo,
            dx,
            minDy
        );
    }

    function flashLoan(
        uint256[] memory amountsOut,
        address to,
        bytes calldata data,
        uint256 deadline
    ) external virtual override whenNotPaused nonReentrant deadlineCheck(deadline) {
        MetaSwapStorage.flashLoan(swapStorage, amountsOut, to, data);
    }

    function addLiquidity(
        uint256[] memory amounts,
        uint256 minMintAmount,
        uint256 deadline
    ) external virtual override whenNotPaused nonReentrant deadlineCheck(deadline) returns (uint256) {
        return MetaSwapStorage.addLiquidity(
            swapStorage,
            metaSwapStorage,
            amounts, 
            minMintAmount
        );
    }

    function removeLiquidity(
        uint256 lpAmount,
        uint256[] memory minAmounts,
        uint256 deadline
    ) external  virtual override nonReentrant deadlineCheck(deadline) returns (uint256[] memory) {
        return MetaSwapStorage.removeLiquidity(swapStorage, lpAmount, minAmounts);
    }

    function removeLiquidityOneToken(
        uint256 lpAmount,
        uint8 index,
        uint256 minAmount,
        uint256 deadline
    ) external virtual override nonReentrant whenNotPaused deadlineCheck(deadline) returns (uint256) {
        return MetaSwapStorage.removeLiquidityOneToken(
            swapStorage, 
            metaSwapStorage, 
            lpAmount, 
            index, 
            minAmount
        );
    }

    function removeLiquidityImbalance(
        uint256[] memory amounts,
        uint256 maxBurnAmount,
        uint256 deadline
    ) external virtual override nonReentrant whenNotPaused deadlineCheck(deadline) returns (uint256) {
        return MetaSwapStorage.removeLiquidityImbalance(
            swapStorage, 
            metaSwapStorage, 
            amounts, 
            maxBurnAmount
        );
    }
}

