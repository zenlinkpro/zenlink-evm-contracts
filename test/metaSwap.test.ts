import chai, { expect } from 'chai'
import { solidity, MockProvider, deployContract } from 'ethereum-waffle'
import { BigNumber, Contract, ContractFactory, Wallet } from 'ethers'
import { ethers } from 'hardhat'

import TestERC20 from '../build/contracts/test/BasicToken.sol/BasicToken.json'
import StableSwap from '../build/contracts/stableswap/StableSwap.sol/StableSwap.json'
import StableSwapStorage from '../build/contracts/stableswap/StableSwapStorage.sol/StableSwapStorage.json'
import MetaSwap from '../build/contracts/stableswap/MetaSwap.sol/MetaSwap.json'
import MetaSwapStorage from '../build/contracts/stableswap/MetaSwapStorage.sol/MetaSwapStorage.json'

import {
  asyncForEach,
  forceAdvanceOneBlock,
  getCurrentBlockTimestamp,
  getUserTokenBalance,
  getUserTokenBalances,
  linkBytecode,
  MAX_UINT256,
  setNextTimestamp,
  setTimestamp,
  TIME,
  ZERO_ADDRESS
} from './shared/utilities'

chai.use(solidity)

const overrides = {
  gasLimit: 6100000
}

describe('MetaSwap', async () => {
  let signers: Array<Wallet>
  let baseSwap: Contract
  let metaSwap: Contract
  let baseLPToken: Contract
  let metaLPToken: Contract
  let dai: Contract
  let usdc: Contract
  let usdt: Contract
  let frax: Contract
  let owner: Wallet
  let user1: Wallet
  let user2: Wallet
  let ownerAddress: string
  let user1Address: string
  let user2Address: string

  // Test Values
  const INITIAL_A_VALUE = 50
  const SWAP_FEE = 1e7
  const LP_TOKEN_NAME = "Test LP Token Name"
  const LP_TOKEN_SYMBOL = "TESTLP"

  const provider = new MockProvider({
    ganacheOptions: {
      hardfork: 'istanbul',
      mnemonic: 'horn horn horn horn horn horn horn horn horn horn horn horn',
      gasLimit: 99999999999,
    },
  })

  async function setupTest() {
    signers = provider.getWallets()
    owner = signers[0]
    user1 = signers[1]
    user2 = signers[2]
    ownerAddress = owner.address
    user1Address = user1.address
    user2Address = user2.address

    dai = await deployContract(
      owner,
      TestERC20,
      ['DAI', 'DAI', '18', '0']
    )
    usdc = await deployContract(
      owner,
      TestERC20,
      ['USDC', 'USDC', '6', '0']
    )
    usdt = await deployContract(
      owner,
      TestERC20,
      ['USDT', 'USDT', '6', '0']
    )
    frax = await deployContract(
      owner,
      TestERC20,
      ['FRAX', 'FRAX', '18', '0']
    )

    const stableSwapStorageContract = await deployContract(owner, StableSwapStorage)
    const stableSwapFactory = (await ethers.getContractFactory(
      StableSwap.abi,
      linkBytecode(StableSwap, {
        'StableSwapStorage': stableSwapStorageContract.address
      }),
      owner,
    )) as ContractFactory
    baseSwap = await stableSwapFactory.deploy()
    
    await baseSwap.initialize(
      [dai.address, usdc.address, usdt.address],
      [18, 6, 6],
      LP_TOKEN_NAME,
      LP_TOKEN_SYMBOL,
      200,
      4e6,
      0,
      owner.address
    )

    baseLPToken = await ethers.getContractAt(
      'LPToken',
      (await baseSwap.swapStorage()).lpToken,
      owner
    )

    // Mint tokens
    await asyncForEach(
      [ownerAddress, user1Address, user2Address],
      async (address) => {
        await dai.setBalance(address, BigNumber.from(10).pow(18).mul(100000))
        await usdc.setBalance(address, BigNumber.from(10).pow(6).mul(100000))
        await usdt.setBalance(address, BigNumber.from(10).pow(6).mul(100000))
        await frax.setBalance(address, BigNumber.from(10).pow(18).mul(100000))
      }
    )

    const metaSwapStorageContract = await deployContract(owner, MetaSwapStorage)
    const metaSwapFactory = (await ethers.getContractFactory(
      MetaSwap.abi,
      linkBytecode(MetaSwap, {
        'StableSwapStorage': stableSwapStorageContract.address,
        'MetaSwapStorage': metaSwapStorageContract.address
      }),
      owner,
    )) as ContractFactory
    metaSwap = await metaSwapFactory.deploy()

    // Set approvals
    await asyncForEach([owner, user1, user2], async (signer) => {
      await frax.connect(signer).approve(metaSwap.address, MAX_UINT256)
      await dai.connect(signer).approve(metaSwap.address, MAX_UINT256)
      await usdc.connect(signer).approve(metaSwap.address, MAX_UINT256)
      await usdt.connect(signer).approve(metaSwap.address, MAX_UINT256)
      await dai.connect(signer).approve(baseSwap.address, MAX_UINT256)
      await usdc.connect(signer).approve(baseSwap.address, MAX_UINT256)
      await usdt.connect(signer).approve(baseSwap.address, MAX_UINT256)
      await baseLPToken.connect(signer).approve(metaSwap.address, MAX_UINT256)

      // Add some liquidity to the base pool
      await baseSwap
        .connect(signer)
        .addLiquidity(
          [String(1e20), String(1e8), String(1e8)],
          0,
          MAX_UINT256,
        )
    })

    // Initialize meta swap pool
    // Manually overload the signature
    await metaSwap.initializeMetaSwap(
      [frax.address, baseLPToken.address],
      [18, 18],
      LP_TOKEN_NAME,
      LP_TOKEN_SYMBOL,
      INITIAL_A_VALUE,
      SWAP_FEE,
      0,
      owner.address,
      baseSwap.address
    )

    metaLPToken = await ethers.getContractAt(
      'LPToken',
      (await metaSwap.swapStorage()).lpToken,
      owner
    )
    
    // Add liquidity to the meta swap pool
    await metaSwap.addLiquidity([String(1e18), String(1e18)], 0, MAX_UINT256)
    
    expect(await frax.balanceOf(metaSwap.address)).to.eq(String(1e18))
    expect(await baseLPToken.balanceOf(metaSwap.address)).to.eq(String(1e18))
    // await metaSwap.addLiquidity([String(1e18), String(1e18)], 0, MAX_UINT256)
  }

  beforeEach(async () => {
    await setupTest()
  })

  describe("swapStorage", () => {
    describe("lpToken", async () => {
      it("Returns correct lpTokenName", async () => {
        expect(await metaLPToken.name()).to.eq(LP_TOKEN_NAME)
      })

      it("Returns correct lpTokenSymbol", async () => {
        expect(await metaLPToken.symbol()).to.eq(LP_TOKEN_SYMBOL)
      })
    })

    describe("A", async () => {
      it("Returns correct A value", async () => {
        expect(await metaSwap.getA()).to.eq(INITIAL_A_VALUE)
        expect(await metaSwap.getAPrecise()).to.eq(INITIAL_A_VALUE * 100)
      })
    })

    describe("fee", async () => {
      it("Returns correct fee value", async () => {
        expect((await metaSwap.swapStorage()).fee).to.eq(SWAP_FEE)
      })
    })

    describe("adminFee", async () => {
      it("Returns correct adminFee value", async () => {
        expect((await metaSwap.swapStorage()).adminFee).to.eq(0)
      })
    })
  })

  describe("getToken", () => {
    it("Returns correct addresses of pooled tokens", async () => {
      expect(await metaSwap.getToken(0)).to.eq(frax.address)
      expect(await metaSwap.getToken(1)).to.eq(baseLPToken.address)
    })

    it("Reverts when index is out of range", async () => {
      await expect(metaSwap.getToken(2)).to.be.reverted
    })
  })

  describe("getTokenIndex", () => {
    it("Returns correct token indexes", async () => {
      expect(await metaSwap.getTokenIndex(frax.address)).to.be.eq(0)
      expect(await metaSwap.getTokenIndex(baseLPToken.address)).to.be.eq(1)
    })

    it("Reverts when token address is not found", async () => {
      await expect(metaSwap.getTokenIndex(ZERO_ADDRESS)).to.be.revertedWith(
        "tokenNotFound",
      )
    })
  })

  describe("getTokenBalance", () => {
    it("Returns correct balances of pooled tokens", async () => {
      expect(await metaSwap.getTokenBalance(0)).to.eq(
        BigNumber.from(String(1e18)),
      )
      expect(await metaSwap.getTokenBalance(1)).to.eq(
        BigNumber.from(String(1e18)),
      )
    })

    it("Reverts when index is out of range", async () => {
      await expect(metaSwap.getTokenBalance(2)).to.be.reverted
    })
  })

  describe("getA", () => {
    it("Returns correct value", async () => {
      expect(await metaSwap.getA()).to.eq(INITIAL_A_VALUE)
    })
  })

  describe("addLiquidity", () => {
    it("Reverts when contract is paused", async () => {
      await metaSwap.pause()

      await expect(
        metaSwap
          .connect(user1)
          .addLiquidity([String(1e18), String(3e18)], 0, MAX_UINT256),
      ).to.be.reverted

      // unpause
      await metaSwap.unpause()
      console.log((await frax.allowance(user1Address, metaSwap.address)).toString())
      console.log((await baseLPToken.allowance(user1Address, metaSwap.address)).toString())

      await metaSwap
        .connect(user1)
        .addLiquidity([String(1e18), String(3e18)], 0, MAX_UINT256)

      const actualPoolTokenAmount = await metaLPToken.balanceOf(user1Address)
      expect(actualPoolTokenAmount).to.eq(BigNumber.from("3991672211258372957"))
    })

    // it("Reverts with 'Amounts must match pooled tokens'", async () => {
    //   await expect(
    //     metaSwap.connect(user1).addLiquidity([String(1e16)], 0, MAX_UINT256),
    //   ).to.be.revertedWith("Amounts must match pooled tokens")
    // })

    // it("Reverts with 'Cannot withdraw more than available'", async () => {
    //   await expect(
    //     metaSwap
    //       .connect(user1)
    //       .calculateTokenAmount([MAX_UINT256, String(3e18)], false),
    //   ).to.be.revertedWith("Cannot withdraw more than available")
    // })

    // it("Reverts with 'Must supply all tokens in pool'", async () => {
    //   metaLPToken.approve(metaSwap.address, String(2e18))
    //   await metaSwap.removeLiquidity(String(2e18), [0, 0], MAX_UINT256)
    //   await expect(
    //     metaSwap
    //       .connect(user1)
    //       .addLiquidity([0, String(3e18)], MAX_UINT256, MAX_UINT256),
    //   ).to.be.revertedWith("Must supply all tokens in pool")
    // })

    // it("Succeeds with expected output amount of pool tokens", async () => {
    //   const calculatedPoolTokenAmount = await metaSwap
    //     .connect(user1)
    //     .calculateTokenAmount([String(1e18), String(3e18)], true)

    //   const calculatedPoolTokenAmountWithSlippage = calculatedPoolTokenAmount
    //     .mul(999)
    //     .div(1000)

    //   await metaSwap
    //     .connect(user1)
    //     .addLiquidity(
    //       [String(1e18), String(3e18)],
    //       calculatedPoolTokenAmountWithSlippage,
    //       MAX_UINT256,
    //     )

    //   const actualPoolTokenAmount = await metaLPToken.balanceOf(user1Address)

    //   // The actual pool token amount is less than 4e18 due to the imbalance of the underlying tokens
    //   expect(actualPoolTokenAmount).to.eq(BigNumber.from("3991672211258372957"))
    // })

    // it("Succeeds with actual pool token amount being within ±0.1% range of calculated pool token", async () => {
    //   const calculatedPoolTokenAmount = await metaSwap
    //     .connect(user1)
    //     .calculateTokenAmount([String(1e18), String(3e18)], true)

    //   const calculatedPoolTokenAmountWithNegativeSlippage =
    //     calculatedPoolTokenAmount.mul(999).div(1000)

    //   const calculatedPoolTokenAmountWithPositiveSlippage =
    //     calculatedPoolTokenAmount.mul(1001).div(1000)

    //   await metaSwap
    //     .connect(user1)
    //     .addLiquidity(
    //       [String(1e18), String(3e18)],
    //       calculatedPoolTokenAmountWithNegativeSlippage,
    //       MAX_UINT256,
    //     )

    //   const actualPoolTokenAmount = await metaLPToken.balanceOf(user1Address)

    //   expect(actualPoolTokenAmount).to.gte(
    //     calculatedPoolTokenAmountWithNegativeSlippage,
    //   )

    //   expect(actualPoolTokenAmount).to.lte(
    //     calculatedPoolTokenAmountWithPositiveSlippage,
    //   )
    // })

    // it("Succeeds with correctly updated tokenBalance after imbalanced deposit", async () => {
    //   await metaSwap
    //     .connect(user1)
    //     .addLiquidity([String(1e18), String(3e18)], 0, MAX_UINT256)

    //   // Check updated token balance
    //   expect(await metaSwap.getTokenBalance(0)).to.eq(
    //     BigNumber.from(String(2e18)),
    //   )
    //   expect(await metaSwap.getTokenBalance(1)).to.eq(
    //     BigNumber.from(String(4e18)),
    //   )
    // })

    // it("Returns correct minted lpToken amount", async () => {
    //   const mintedAmount = await metaSwap
    //     .connect(user1)
    //     .callStatic.addLiquidity([String(1e18), String(2e18)], 0, MAX_UINT256)

    //   expect(mintedAmount).to.eq("2997459774673651937")
    // })

    // it("Reverts when minToMint is not reached due to front running", async () => {
    //   const calculatedLPTokenAmount = await metaSwap
    //     .connect(user1)
    //     .calculateTokenAmount([String(1e18), String(3e18)], true)

    //   const calculatedLPTokenAmountWithSlippage = calculatedLPTokenAmount
    //     .mul(999)
    //     .div(1000)

    //   // Someone else deposits thus front running user 1's deposit
    //   await metaSwap.addLiquidity([String(1e18), String(3e18)], 0, MAX_UINT256)

    //   await expect(
    //     metaSwap
    //       .connect(user1)
    //       .addLiquidity(
    //         [String(1e18), String(3e18)],
    //         calculatedLPTokenAmountWithSlippage,
    //         MAX_UINT256,
    //       ),
    //   ).to.be.reverted
    // })

    // it("Reverts when block is mined after deadline", async () => {
    //   const currentTimestamp = await getCurrentBlockTimestamp(provider)
    //   await setNextTimestamp(provider, currentTimestamp + 60 * 10)

    //   await expect(
    //     metaSwap
    //       .connect(user1)
    //       .addLiquidity(
    //         [String(2e18), String(1e16)],
    //         0,
    //         currentTimestamp + 60 * 5,
    //       ),
    //   ).to.be.revertedWith("Deadline not met")
    // })

    // it("Emits addLiquidity event", async () => {
    //   const calculatedLPTokenAmount = await metaSwap
    //     .connect(user1)
    //     .calculateTokenAmount([String(2e18), String(1e16)], true)

    //   const calculatedLPTokenAmountWithSlippage = calculatedLPTokenAmount
    //     .mul(999)
    //     .div(1000)

    //   await expect(
    //     metaSwap
    //       .connect(user1)
    //       .addLiquidity(
    //         [String(2e18), String(1e16)],
    //         calculatedLPTokenAmountWithSlippage,
    //         MAX_UINT256,
    //       ),
    //   ).to.emit(metaSwap.connect(user1), "AddLiquidity")
    // })
  })
})
