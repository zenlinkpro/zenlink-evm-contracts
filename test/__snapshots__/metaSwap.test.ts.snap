// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetaSwap swapUnderlying Succeeds with expected swap amounts From 18 decimal token (meta) to 18 decimal token (base) 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 240789,
}
`;
